### PR TITLE
CFI: Fix many vtable-related problems

### PR DESCRIPTION
**Pull request by maurer**
This PR makes it possible to use trait objects with CFI enabled, as checked by the various tests added.

Non-landable tests I also did:
* Before I rebased (and was on a version of LLVM that matched my system LLVM), I was able to run `./x.py build --stage 2` with CFI hacked on for all build targets after `#[no_sanitize]`ing parts of the `proc-macro` crate. This gives me some confidence in the completeness of these CFI modifications.
* I am able to build and run a KCFI'd Rust-language driver in the Linux kernel, and use it to boot Android:
```
mmaurer@anyblade:~/android/platform$ adb shell cat /proc/config.gz | gunzip | grep 'BINDER\|CFI'
CONFIG_FUNCTION_PADDING_CFI=11
CONFIG_ARCH_SUPPORTS_CFI_CLANG=y
CONFIG_ARCH_USES_CFI_TRAPS=y
CONFIG_CFI_CLANG=y
# CONFIG_CFI_PERMISSIVE is not set
# CONFIG_SND_SOC_FSL_MICFIL is not set
# CONFIG_ANDROID_BINDER_IPC is not set
CONFIG_ANDROID_BINDER_IPC_RUST=y
CONFIG_ANDROID_BINDERFS_RUST=y
CONFIG_ANDROID_BINDER_DEVICES_RUST="binder,hwbinder,vndbinder"
mmaurer@anyblade:~/android/platform$
```

The general approach taken here is is to make `Virtual` instances and the instances put into vtables match their abstracted type, without changing the type of the underlying function instances.

cc @rcvalle 
cc @tmandry 
cc @workingjubilee 

I'm not sure who an appropriate reviewer is for this, so I'll let rustbot take a guess and post this PR in the zulip thread.

### CODE REVIEW

**Comment by petrochenkov** [comment]
>I'm not sure who an appropriate reviewer is for this, so I'll let rustbot take a guess and post this PR in the zulip thread.

Looks like rustbot chose me due to the `rustc_target` changes, but I'm not an appropriate reviewer here.
r? compiler

**Comment by workingjubilee** [comment]
@maurer Is there no chance at factoring this down the line of the `mir -> cg_ssa` change?

Why does this require changes in `rustc_const_eval`?

File: compiler/rustc_middle/src/ty/instance.rs [hunk 1]
```diff
@@ -135,14 +135,27 @@ pub enum InstanceDef<'tcx> {
...
     FnPtrAddrShim(DefId, Ty<'tcx>),
+
+    /// Typecast shim which replaces the `Self` type with the provided type.
+    /// This is used in vtable calls, where the type of `Self` is abstract as of the time of
+    /// the call.
+    ///
+    /// For `target_instance` which are generatable shims, this is done via inserting a cast at the
+    /// beginning of the shim generated by that instance.
+    /// For `Item`s, we instead build a direct call shim with that cast inserted.
+    CfiShim { target_instance: &'tcx InstanceDef<'tcx>, invoke_ty: Ty<'tcx> },
```
Comment (RalfJung) [inline compiler/rustc_middle/src/ty/instance.rs:146]: Does this mean we can remove the code in the interpreter and codegen that currently does this type adjustments? In the interpreter, that's [here](https://github.com/rust-lang/rust/blob/fc9d1a8133616c4cd6db7c2bcf9f4981e8860be5/compiler/rustc_const_eval/src/interpret/terminator.rs#L857-L872).

If that code is needed then I don't understand what this achieves.

**Comment by maurer** [comment]
> @maurer Is there no chance at factoring this down the line of the `mir -> cg_ssa` changeover?

I'm unsure how that would work - if I have
```
impl MyTrait for Foo {
   fn bar(&self) { // ... }
}
trait OtherTrait : MyTrait {}
impl OtherTrait for Foo {}
fn main() {
  Foo.bar(); // Base instance, no shims
  let x = &Foo as &dyn MyTrait;
  x.bar(); // Shimmed instance, new invocation type `dyn MyTrait`
  let y = &Foo as &dyn OtherTrait;
  y.bar(); // Shimmed instance, new invocation type `dyn OtherTrait`
}
```

how could I introduce this cutover at the stage of MIR->SSA? I was under the impression that by the time that was occurring, we were supposed to be doing a 1:1 transformation of Instance -> codegen structure, not Instance -> several. VTable generation has also already occurred above that abstraction level, so I don't know how I would get the vtables to point to each of the three specialized allocations.

File: compiler/rustc_middle/src/ty/instance.rs (hunk 1 above)
Comment (maurer) [reply compiler/rustc_middle/src/ty/instance.rs:146]: That code is still needed at the moment because I am conditionally generating shims - they are only generated when CFI is enabled, since they will result in worse code generation (If it's an item, there's a cast trampoline, and if it's not an item, there's a variant of the shim produced). Even if all calls were inlined, it'd still produce extra bulk as each vtable the type appears in gets its own shim.

I could try testing disabling that code when `.cfi_shims()` is true if you think that would be interesting.

**Comment by RalfJung** [comment]
So this makes the two vtables point to two different functions for that example? That sounds pretty bad for code size, doesn't it? And for perf too -- either there's now extra calls and stack traffic, or it gets inlined and there's even more code bloat and the same code has to be optimized and passed to the backend multiple times.

**Comment by workingjubilee** [comment]
> how could I introduce this cutover at the stage of MIR->SSA? I was under the impression that by the time that was occurring, we were supposed to be doing a 1:1 transformation of Instance -> codegen structure, not Instance -> several. VTable generation has also already occurred above that abstraction level, so I don't know how I would get the vtables to point to each of the three specialized allocations.

I only meant in terms of landing the changes that need to happen before `cg_ssa` in a separate PR, so that they can be reviewed separately. But if that isn't very realistic and not very useful in your opinion, then don't worry about it: we have bigger fish to fry. It just is common for knowledge of the compiler to be split on that bifurcating point, so this patch will likely require multiple reviewers.

File: compiler/rustc_middle/src/ty/instance.rs (hunk 1 above)
Comment (maurer) [reply compiler/rustc_middle/src/ty/instance.rs:146]: Actually, my adjustment to `Virtual` (which is unconditional) might make that work, I'll try it in a bit.

**Comment by maurer** [comment]
> It adds a new kind of MIR shim. But the question then is, why is that needed? @maurer please provide a description of your design and the reasons for it. Do not expect the reader to know how CFI works, so you may want to start with a problem statement that explains the relevant parts of CFI. This sounds like it adds an extra function indirection to virtual calls, did you benchmark that?

I'll provide a larger writeup in a bit, but here's the summary:

* CFI works by checking the type of the function pointer it expects to call before it calls it, and making sure it matches. These function pointer types are annotated to LLVM, and it provides the facility to test the type.
* Rust vtables today intentionally have an implicit cast at the site of any indirect call. This is why you see the thin pointer logic you pointed to - you're implicitly casting a `*dyn MyTrait` to a `&Foo` while discarding the vtable so that the caller's expected ABI matches the callee's expected ABI, even if the signatures don't match.
* This patch makes it so that rather than taking a `&Foo`, vtable entries will take a `*dyn MyTrait`. It does this by creating an additional shim for every entry, either grafted on for generated code or via a call trampoline otherwise, which has a MIR cast from the general type to the specific type.

With regards to performance, I have not benchmarked it, though I do expect it to hurt performance. Since most of these methods would previously crash when CFI was enabled, and the shims are only generated with CFI enabled, I didn't think it was critical.

If you'd like, I can benchmark enabling shims *without* enabling CFI, do you have a suggested workload?

**Comment by maurer** [comment]
> > how could I introduce this cutover at the stage of MIR->SSA? I was under the impression that by the time that was occurring, we were supposed to be doing a 1:1 transformation of Instance -> codegen structure, not Instance -> several. VTable generation has also already occurred above that abstraction level, so I don't know how I would get the vtables to point to each of the three specialized allocations.
> 
> I only meant in terms of landing the changes that need to happen before `cg_ssa` in a separate PR, so that they can be reviewed separately. But if that isn't very realistic and not very useful in your opinion, then don't worry about it: we have bigger fish to fry. It just is common for knowledge of the compiler to be split on that bifurcating point, so this patch will likely require multiple reviewers.

Sorry, I misunderstood and thought you wanted me to push all the logic into `cg_ssa` or beyond. I could probably cut the PR at [CFI: Apply CFI shims to drops](https://github.com/rust-lang/rust/pull/121962/commits/ec3dbae0c0216b1cd7af740ae7ab03eafe17743a), and land each patch beyond in individual PRs, it's only the initial block that kind of all needs to go in at once to do anything useful/meaningful. Would that be helpful? Last I checked there wasn't a good way to stack PRs on github, but I could file separate ones at each reviewable point if that would help people work through things.

**Comment by RalfJung** [comment]
> I'll provide a larger writeup in a bit, but here's the summary:

Seems unfortunate that CFI doesn't have a better way to encode the correct semantics here and requires some contortions in the generated code. But well, I guess it's what it is...

> If you'd like, I can benchmark enabling shims without enabling CFI, do you have a suggested workload?

I didn't realize this would be enabled only for CFI, that makes perf less of a concern then.

It introduces new risks though, if these codepaths are not executed in regular builds at all. I see you added some tests :) . I would encourage you to not unnecessarily split up tests into multiple files; the overhead of separate compiler invocations does add up across a large test suite like ours. (But if they need different flags then they do need different files.)

**Comment by workingjubilee** [comment]
> CFI works by checking the type of the function pointer it expects to call before it calls it, and making sure it matches. These function pointer types are annotated to LLVM, and it provides the facility to test the type.

Hmm. Why exactly is the KCFI model incompatible with this particular style of pointer cast? My understanding is that usually, other kinds of pointer-casting can be accepted, without necessarily erasing all type-checking entropy, and that in many cases, C code that needs to do similar will be performing such an erasure manually?

**Comment by maurer** [comment]
> So this makes the two vtables point to two different functions for that example? That sounds pretty bad for code size, doesn't it? And for perf too -- either there's now extra calls and stack traffic, or it gets inlined and there's even more code bloat and the same code has to be optimized and passed to the backend multiple times.
> 
Yes, they point to two different functions. This is one way to make it work for regular CFI, and mandatory for KCFI where each function is only allowed a single type. We might be able to get rid of shims some day for regular CFI by leveraging the ability to associate multiple types with a function, but KCFI will always need these shims. KCFI might be able to use only *one* shim if we make trait upcasting a load-bearing part of virtual calls (since then you'd only ever shim the function at the trait it was implemented for, not all child traits), but that would be future work.

> I guess I'll wait for your design document. Currently that doesn't sound like somewhat we want to do for regular builds. Maybe this only happens for CFI builds, the PR description doesn't say.
I don't think we want to generate these cast shims for regular builds. The only change that I made that currently affects regular builds is the last one on the stack which strips auto traits off the receivers of virtual calls.

**Comment by maurer** [comment]
> > CFI works by checking the type of the function pointer it expects to call before it calls it, and making sure it matches. These function pointer types are annotated to LLVM, and it provides the facility to test the type.
> 
> Hmm. Why exactly is the KCFI model incompatible with this particular style of pointer cast? My understanding is that usually, other kinds of pointer-casting can be accepted, without necessarily erasing all type-checking entropy, and that in many cases, C code that needs to do similar will be performing such an erasure manually?

KCFI only supports exactly one principal type for a function. This means that no casting at the indirect call site is allowed. The actual implementation mechanism is that the compiler takes the principal type, encodes it to a string, hashes it, and then puts the hashed type in a prelude bundle before the function. When a caller wants to call a function, the compiler takes the type, encodes it to a string, hashes it, and then inserts a check that the function has that hashed type right before it.

In the one place where KCFI is heavily used, Linux, C code *does* indeed perform this erasure manually. Since they don't have objects, their equivalent of vtables are usually structs of function pointers. Their usual pattern is that your implementation takes a pointer to a struct of the abstract type, which is sometimes data-bearing. If it is not data-bearing, it will usually just be a void* you do an explicit cast to your real object on. If it is, then usually it is present as a field inside your object, and the `container_of` macro is used to do pointer math and a cast on it, assuming that it is a pointer into your object.

It is possible to avoid these shims for LLVM-CFI, but that requires getting the C++ subtyping relation to represent the Rust legal-casting relation.

**Comment by workingjubilee** [comment]
> 
> Sorry, I misunderstood and thought you wanted me to push all the logic into `cg_ssa` or beyond. I could probably cut the PR at [CFI: Apply CFI shims to drops](https://github.com/rust-lang/rust/pull/121962/commits/ec3dbae0c0216b1cd7af740ae7ab03eafe17743a), and land each patch beyond in individual PRs, it's only the initial block that kind of all needs to go in at once to do anything useful/meaningful. Would that be helpful? Last I checked there wasn't a good way to stack PRs on github, but I could file separate ones at each reviewable point if that would help people work through things.

Possibly. You're right about stacked diffs not being a thing on GH. I'll wait for someone from mir-opt to make a noise about whether they think also-seeing the LLVM and symbol-mangling logic is helpful.

Although, now that I look more closely at it, could you possibly pluck these out? They don't seem like they depend on the rest of the changes at all, though I could be mistaken. Those could be landed in their own PR, if so.
- ed48012068b230cf528b69d3a8507a3e90448eee
- 415e2c4b219ee34043722f435219b0f51a5edd7d

**Comment by rcvalle** [comment]
Is it possible to break these changes into a set of smaller self-contained pairs of issues and fixes (like I've been fixing the CFI known issues)? The reason I ask is that not only it will make it easier for me to understand the problems and what is being proposed, but also because I think some of the issues here are fixed already or are in the process of being fixed for CFI using type metadata IDs/encoding only.

I understand that CFI and KCFI perform type tests differently and are implemented as different LLVM passes for KCFI to not require LTO, and because of that KCFI has the limitation that a function may have one type id assigned only, and thus why shims *may* be necessary. However, CFI support for Rust was designed to be as unobtrusive as possible by just generating type tests and using type metadata IDs/encoding for its implementation, and I'd like to keep it that way (at least for CFI, if not possible for KCFI).

Introducing shims/trampolines may result in CFI bypasses and affect CFI performance and I'd like to avoid that. Therefore, would it be also possible to (in addition to breaking these changes into a set of smaller self-contained pairs of issues and fixes), also making them for KCFI only (and not for CFI)?

I'm not sure (and others let me know what they think), but I think that--even for KCFI only--this may require at least a design doc and/or an MCP explaining what is being proposed in detail--it'd for sure help me better understand and review it.

Have you investigated adding support for functions to have multiple type ids assigned to the LLVM passes for KCFI instead, and weighted the costs x benefits of doing that instead of implementing a workaround for it in the Rust compiler? I think it's important it's also documented there.

**Comment by workingjubilee** [comment]
Despite my normally laissez-faire attitude about encoding certain type relationships, I strongly agree that we wish to handle generating code that produces _more_ gadgets that are essentially explicitly designed to allow breaking down restrictions on type encodings with extra care.

**Comment by maurer** [comment]
> In the case of
> 
> ```rust
> impl MyTrait for Foo {
>    fn bar(&self) { // ... }
> }
> trait OtherTrait : MyTrait {}
> impl OtherTrait for Foo {}
> ```
> 
> can't we generate the `<Foo as MyTrait>::bar` instance with `*const MyTrait` as self type even without the shim? For a `<Foo as OtherTrait>::bar` call, we can still use `*const MyTrait` as self type, so that doesn't need a shim either.

This isn't true, see the address-taken test I added. I can elaborate more when I'm not stuck on my phone if that doesn't explain things.

**Comment by bjorn3** [comment]
> This isn't true, see the address-taken test I added.

That is a cast to a function pointer, right? See the rest of my comment:

> It may require a shim for `<Foo as MyTrait>::bar as fn(&Foo)`, but we already have `InstanceDef::ReifyShim` for that (it is technically for `<dyn MyTrait as MyTrait>::bar`, but I don't see why it wouldn't work for the concrete case either).

**Comment by maurer** [comment]
> > It may require a shim for `<Foo as MyTrait>::bar as fn(&Foo)`, but we already have `InstanceDef::ReifyShim` for that (it is technically for `<dyn MyTrait as MyTrait>::bar`, but I don't see why it wouldn't work for the concrete case either).

Sorry, missed that part - I think `ReifyShim` would work there with appropriate additional type adjustments (it'd need to know how to go back to the original def_id's typesig before the `*const MyTrait` rewrite).

However would we get `<dyn MyTrait as MySuperTrait>::bar` to have receiver `*dyn MySuperTrait`? Specifically:
```
trait MyTrait {
    fn foo(&self);
}

trait MySubTrait: MyTrait {}

struct Foo;

impl MyTrait for Foo {
    fn foo(&self) {}
}

impl MySubTrait for Foo {}

fn main() {
    let f: fn(_) = <dyn MySubTrait>::foo;
    println!("{}", std::any::type_name_of_val(&f))
}
```
will print `fn(&dyn playground::MySubTrait)`.

The problem here is that if what `ReifyShim` does is to swap `*const dyn MyTrait` to `&Foo` by examining the type of the original function, we still need a shim kind to rewrite `*const dyn MyTrait` to `*const dyn MySubTrait` - this is exactly what the `CfiShim` does, it lets you select what the self type needs to be rewritten to.

I do think you have a point here that if we make `*const dyn MyTrait` the *default* for trait methods, and require a shim for instances where a function pointer is created, that would likely result in fewer shims as I don't think they're common in code.

Additionally, to make this work we'd need to switch the receiver type on `Virtual` instances based on which trait defined the item inside, but that could be doable - otherwise we'd still need a separate shim per vtable.

I'm still writing up a design summary, but if you have an approach that avoids introducing a new shim type, that would be pretty cool and I could try to rework the code.

**Comment by maurer** [comment]
The [promised design doc](https://hackmd.io/6U7gB9SkQBWV9l4iIsp3Jg)!

**Comment by maurer** [comment]
The [promised design doc](https://hackmd.io/6U7gB9SkQBWV9l4iIsp3Jg)!

**Comment by rcvalle** [comment]
Thanks for the design doc! From the design doc, it seems these are the issues you're fixing with what is being proposed:

1. `<S as Foo>::foo as fn(&S)` where `Foo` is a trait with `fn foo(&self)` - you can't use methods from traits as functions
2. Calling a method with signature `fn foo(self: Arc<Self>)` on a trait object - you can't use anything but `&self` or `&mut self` as a receiver
3. `let f: &fn() = &((|| ()) as _);` - you can't convert a closure to a callable function pointer[^fixinprogress]
4. `let _ = Box<dyn Foo> = Box::new(S);` - you can't drop a trait object, the drop entry has the wrong type.
5. If you use `self_cell`, you'll get an infinite loop

Is that correct? If so, I'd recommend listing these cases at the introduction because from it, it isn't clear what exactly the issues are.

From these issues, (1) and (3) are known issues and I'm currently working on a fix for (3) for both CFI and KCFI using type metadata IDs/encoding only (see #116404), and programs that don't use any of these patterns listed work, and with this fix also rebuilding the standard library using the Cargo build-std feature--and all the standard library tests also pass.

The other issues seem like they can be fixed for both CFI and KCFI also using type metadata IDs/encoding only, or using the solution I'm also currently working on and I mentioned to you earlier (and I see you also listed it in the design doc) of attaching two type ids to trait methods, one using the identity of the trait that implements the method (normalized during encoding) and the other using the concrete self.

This could also work for KCFI if instead of reading a 32-bit hash and testing it, it read a 64-bit hash, tested the first half of it first, which would be the hash of first type id and the most likely to pass, and if it failed tested the second half of it, which would be the second type id and the less common case. I think this would be simpler/less risky and more efficient than using shims/trampolines (but I have suggestions for possibly better solutions below). This could be worse for 32-bit architectures because it'd require two loads, but I think it'd would still be better than the double indirection and the possible cache coherence impact of the shim/trampoline being placed farther away from the actual destination.

Have you investigated it or equivalent solution? I don't see anything in the design doc. I still think we should investigate adding support for functions to have multiple type ids assigned to the LLVM passes for KCFI, and weighting the costs x benefits of doing that instead of implementing a workaround for it in the Rust compiler.

Alternatively, there are a few possible improvements for the LLVM pass for CFI that I want to explore in the future that may be a possible solution for KCFI if also implemented in the LLVM passes for KCFI (i.e., support for wildcards and basic expressions in type ids and types):

1. Basic type id expressions (i.e., an "or" expression):
   Even though KCFI has the limitation that a function may have one type id assigned only, the LLVM passes for KCFI could support a basic type id expressions, such as `"(typeid1|typeid2)"` (with its actual syntax to be defined). (E.g.: `"_ZTSFviE"`, `"_ZTSFvu3i32E"`, and `"(_ZTSFviE|_ZTSFvu3i32E)"`.) For KCFI, `typeid1`, `typeid2`, and the type id expression would resolve to the same unique id `uniqueid1` and it would use the `xxHash` of this unique id. This has the disadvantage that functions having the `typeid1`, `typeid2`, or `"(typeid1|typeid2)"` assigned would be coalesced into the same type (but this would likely be used for functions that needed to be compatible anyway but couldn't be encoded correctly because of information loss). For CFI, the function with the expression assigned could just have both `typeid1` and `typeid2` assigned instead.

2. Basic type expressions (i.e., an "or" expression):
   Similarly to type id expressions, the LLVM passes for KCFI could support basic type expressions, such as `"(type1|type2)"` within the full type id (with its actual syntax to be defined). (E.g.: `"_ZTSFviE"`, `"_ZTSFvu3i32E"`, and `"_ZTSFv(i|u3i32)E)"`.)

These solutions remove the burden from the Rust compiler resulting from of a limitation of the LLVM passes for KCFI, and may solve the problem in a way that may also be useful for other languages without also placing the burden on their compilers.

**Comment by maurer** [comment]
> Thanks for the design doc! From the design doc, it seems these are the issues you're fixing with what is being proposed:
> 
> 1. `<S as Foo>::foo as fn(&S)` where `Foo` is a trait with `fn foo(&self)` - you can't use methods from traits as functions
> 2. Calling a method with signature `fn foo(self: Arc<Self>)` on a trait object - you can't use anything but `&self` or `&mut self` as a receiver
> 3. `let f: &fn() = &((|| ()) as _);` - you can't convert a closure to a callable function pointer[^fixinprogress]
> 4. `let _ = Box<dyn Foo> = Box::new(S);` - you can't drop a trait object, the drop entry has the wrong type.
> 5. If you use `self_cell`, you'll get an infinite loop
> 
> Is that correct? If so, I'd recommend listing these cases at the introduction because from it, it isn't clear what exactly the issues are.

This is correct, but the trick isn't just fixing them in some cases, it's figuring out how to fix them in all cases without breaking each other. For example, the previous fix for calling trait objects broke address-taken. There are several simpler "fixes" for `drop_in_place` that could be written that would fail for some trait hierarchies.

> and programs that don't use any of these patterns listed work
Using trait objects that drop is not a rare pattern. Using alternate receivers is load bearing for the kernel.
> and with this fix also rebuilding the standard library using the Cargo build-std feature--and all the standard library tests also pass.

My solution can also build and run the whole *compiler* with CFI enabled.

> The other issues seem like they can be fixed for both CFI and KCFI also using type metadata IDs/encoding only, or using the solution I'm also currently working on and I mentioned to you earlier (and I see you also listed it in the design doc) of attaching two type ids to trait methods, one using the identity of the trait that implements the method (normalized during encoding) and the other using the concrete self.
If you mean avoiding having the shims change the type, I agree that this is possible, and it's mentioned in the design doc and the tradeoffs are discussed. If you mean having no shims, I spend an entire section in the doc explaining that:
1. As it stands, the information you need to normalize the receiver is erased from the type before you get it. (We could still compute from the `Instance`).
2. Having no shims cannot support KCFI due to supporting a single attachment.
> 
> This could also work for KCFI if instead of reading a 32-bit hash and testing it, it read a 64-bit hash, tested the first half of it first, which would be the hash of first type id and the most likely to pass, and if it failed tested the second half of it, which would be the second type id and the less common case. I think this would be simpler/less risky and more efficient than using shims/trampolines (but I have suggestions for possibly better solutions below). This could be worse for 32-bit architectures because it'd require two loads, but I think it'd would still be better than the double indirection and the possible cache coherence impact of the shim/trampoline being placed farther away from the actual destination.
> 
> Have you investigated it or equivalent solution? I don't see anything in the design doc. I still think we should investigate adding support for functions to have multiple type ids assigned to the LLVM passes for KCFI, and weighting the costs x benefits of doing that instead of implementing a workaround for it in the Rust compiler.

You're proposing declaring that we can't support KCFI, creating an alternative, ABI-breaking mechanism KCFI 2.0, patching LLVM to support it, then patching the kernel to support it (including its runtime self-patching, which also adjusts KCFI in various ways). Doing this would delay our ability to use this feature by more than a year due to pipelines and time spent convincing LLVM and the kernel - I'm sure you experienced some of how long this can take to do when trying to add the generalize integers option.

In short, it isn't in the design doc because you're not suggesting a way to add support for KCFI - you're suggesting a way to change KCFI to make it easier to support.

**Comment by RalfJung** [comment]
Thanks for the extensive writeup!

> This led to the [introduction](https://reviews.llvm.org/D119296) of KCFI. This makes a few observations:
>-   C (not C++) lacks vtables to protect or typesign
>-    All executable memory should be read-only
>-    C doesn't have polymorphism, so every function has a principal type

So that sounds like a CFI mechanism developed specifically for C, which is much less expressive than Rust, and now we're trying to shove Rust into this C-shaped hole. I guess I am not surprised that there is some friction.^^

But if there's a way to do that without too much maintenance burden for rustc and without a big risk of accidentally producing broken code, then sure, why not. :)  (Also I'm not on the compiler team so it's not up to me anyway.)

**Comment by workingjubilee** [comment]
ah, another thriving day as a member of T-cursed

**Comment by maurer** [comment]
I'll split those off in a moment then. I may have to strip tests off of them because they may depend on other changes in the stack, but the semantics changes can go in alone.

If you're going to be reviewing it, do you think it'd be worth me doing an alternate implementation following the [Separate Machinery](https://hackmd.io/6U7gB9SkQBWV9l4iIsp3Jg#Separate-Machinery) approach from the design document, or should I hold off until you've had a chance to look at it as is?

**Comment by maurer** [comment]
In that model, the shims still get added, just as they do now. The primary differences are:
1. The `Instance` will still get passed to the CFI encoder, not the type
2. The shims will no longer have a MIR adjustment - they will generate identical-but-non-mergeable code.
3. We might even need to pass the `Instance` to the CFI encoder for virtual calls as well, I'll have to try it to be certain.

There will still be a `CfiShim` type, both queries I added will still be required, and the `.cfi_shim()` method where most of the magic happens will still be needed, but the changes in `rustc_mir_transform` would become limited to "If it's a CFI shim, just generate the wrapped instance", and the `Instance::ty` function would remain unchanged.

Because the `cfi_shim` method occasionally rewrites the original instance to a `Reify` with different arguments, different MIR will be generated sometimes, but no modifications to the shim generator would be necessary anymore, and the compiler-side types of vtable entries wouldn't change anymore.

**Comment by workingjubilee** [comment]
Hmm.

In general I would prefer to see this change that we discussed earlier as an "alternate take":

> I do think you have a point here that if we make `*const dyn MyTrait` the default for trait methods, and require a shim for instances where a function pointer is created, that would likely result in fewer shims as I don't think they're common in code.

And skimming the doc again, it seems the Separate Machinery proposal isn't that, but rather a redef of the alias set, right?

**Comment by maurer** [comment]
> Hmm.
> 
> In general I would prefer to see this change that we discussed earlier as an "alternate take":
> 
> > I do think you have a point here that if we make `*const dyn MyTrait` the default for trait methods, and require a shim for instances where a function pointer is created, that would likely result in fewer shims as I don't think they're common in code.
I'm not convinced this is easy, but if you think exploration of it is a prereq for landing rather than a follow-on, I can give it a shot.
> 
> And skimming the doc again, it seems the Separate Machinery proposal isn't that, but rather a redef of the alias set, right?
No, same alias set, just changing where and how it's computed in the compiler. The current implementation makes the shimmed functions types line up with what the caller expects it to be. This implementation would leave the type at the underlying type, but make the shimmed alias set *only* line up with what the caller expects them to be.

The output object code in each case would be identical, it's just a rework of where the changes are in the compiler to reduce changes in `mir_transform` and `middle`.

**Comment by workingjubilee** [comment]
I see, I see. Yes, re: the `*const dyn Trait` change: If you try hacking on it for a few hours and it turns out to be a huge bear to implement, that would be valuable information. I am sure @bjorn3 would be happy to hash out the idea a bit more with you, too. It just seems like it makes a few things easier on everyone.

But I think I would like to see Separate Machinery first, then, as it seems significantly more different than this, and I have a stronger impression that I won't really be able to judge without seeing it.

**Comment by rcvalle** [comment]
I'd also like to reiterate my request for, in addition to breaking these changes into a set of smaller self-contained pairs of issues and fixes, which seems you started to do, also making these changes for KCFI only (since they aren't needed for CFI, and I'm fixing those corner cases for CFI using a solution that doesn't require any shims). 

In general, I'd like to see what the absolute minimum necessary use of shims even for KCFI looks like. For example, for function items, closures, and Fn trait objects to be used interchangeably, there is a solution I mentioned there I'm working on (see # 116404) by signature_unclosure'ing/transforming those, exactly as the compiler does at the call sites and CFI does for other equivalent types that works even for KCFI, so the shims aren't necessary in this case even for KCFI. Therefore, if you also wait these changes to be merged, less shims are necessary.

Since KCFI landed pretty recently (for the Linux kernel standards), I'm still not convinced that it is too widely deployed yet  that makes it very difficult to explore any changes to it. There are also the points I made about the possibility of inadvertently introducing KCFI bypasses, the possibility of general performance and cache coherence/locality impact of the shim/trampoline being placed farther away from the actual destination, that I think need to be looked at, and more generally what is the right thing to do longer term, both for the Rust project and the Linux kernel. After all, there is a reason the CFI project is taking long, and it's because we've been trying to do the right thing, and not rush things--if we wanted, we could have everything building and running already, but would that have been the right thing to do?

**Comment by rcvalle** [comment]
For anyone that commented on this PR and also anyone new to this PR, see my comments in https://github.com/rust-lang/rust/pull/116404#issuecomment-2008602437.

**Comment by workingjubilee** [comment]
Well, this obviously would need a rebase, but also it's not clear that we're going to take exactly this route since #123012 so!

@rustbot author

**Comment by maurer** [comment]
I'm closing this PR as I'm taking an approach much closer to the type removal one. The branch will still be there for reference until we finish fixing the overall vtable issues in case we need to crib solutions from it.
